File: lib/ios/DotIndicatorOptions.m
#import "DotIndicatorOptions.h"
#import "RNNNullBool.h"
#import "RNNNullColor.h"
#import "RNNNullNumber.h"

@implementation DotIndicatorOptions
- (instancetype)initWithDict:(NSDictionary *)dict {
    self = [super initWithDict:dict];

    self.color = [ColorParser parse:dict key:@"color"];
    self.size = [NumberParser parse:dict key:@"size"];
    self.visible = [BoolParser parse:dict key:@"visible"];
    return self;
}

- (instancetype)init {
    _color = [RNNNullColor new];
    _size = [RNNNullNumber new];
    _visible = [RNNNullBool new];
    return self;
}

- (void)mergeOptions:(DotIndicatorOptions *)options {
    if (options.color.hasValue)
        self.color = options.color;
    if (options.size.hasValue)
        self.size = options.size;
    if (options.visible.hasValue)
        self.visible = options.visible;
}

- (bool)hasValue {
    return [self.visible hasValue];
}

@end

File: lib/android/app/src/main/java/com/reactnativenavigation/options/DotIndicatorOptions.java
package com.reactnativenavigation.options;

import android.content.Context;

import androidx.annotation.Nullable;

import com.reactnativenavigation.options.params.Bool;
import com.reactnativenavigation.options.params.NullBool;
import com.reactnativenavigation.options.params.NullNumber;
import com.reactnativenavigation.options.params.Number;
import com.reactnativenavigation.options.params.RNNColour;
import com.reactnativenavigation.options.params.RNNColourKt;
import com.reactnativenavigation.options.params.RNNNullColor;
import com.reactnativenavigation.options.parsers.BoolParser;
import com.reactnativenavigation.options.parsers.NumberParser;

import org.json.JSONObject;

public class DotIndicatorOptions {
    public static DotIndicatorOptions parse(Context context, @Nullable JSONObject json) {
        DotIndicatorOptions options = new DotIndicatorOptions();
        if (json == null) return options;

        options.color = RNNColourKt.parse(context, json.optJSONObject("color"));
        options.size = NumberParser.parse(json, "size");
        options.visible = BoolParser.parse(json, "visible");
        options.animate = BoolParser.parse(json, "animate");

        return options;
    }

    public RNNColour color = new RNNNullColor();
    public Number size = new NullNumber();
    public Bool visible = new NullBool();
    public Bool animate = new NullBool();

    public boolean hasValue() {
        return visible.hasValue();
    }
}
